### ISSUE TITLE
User is not able to receive messages with Signal 7.20.2 even after upgrading to Signal 7.22.1

### ISSUE THREAD

**Issue by signal-support-request** [signalapp/Signal-Android#13753]
### Guidelines

- [X] I have searched searched open and closed issues for duplicates
- [X] I am submitting a bug report for existing functionality that does not work as intended
- [X] This isn't a feature request or a discussion topic

### Bug description

User of Signal 7.20.2 on Android is not able to receive any messages, even after upgrading to Signal 7.22.1 from Google Play version to manually downloaded https://github.com/signalapp/Signal-Android/releases/download/v7.22.1/Signal-Android-website-prod-universal-release-7.22.1.apk
Sending of read receipts and actual messages apparently also does not happen even though the logs indicate successful sending of some messages like sender keys when user is added to the group. But no new messages are visible in the UI. There is a notification "You may have new messages" stuck at the top and "Checking for messages..." with a progress bar below it, so apparently FCM notification arrived but Signal fails to fetch actual messages.

Linked Desktop works fine at the same time. Another user on the same Wi-Fi using Android also has no problems, so it is not a network connectivity issue.

There is a similar issue saying that receiving works and sending does not: https://github.com/signalapp/Signal-Android/issues/13750

There was a similar closed issue at https://github.com/signalapp/Signal-Android/issues/13726 that was closed as a duplicate of https://github.com/signalapp/Signal-Android/issues/13724
It was attributed to some slow SQL migration, but now the version is 7.22.1 and there is no indication of any slow migration in the logs, all migrations finished quickly.

### Device

Samsung Galaxy Tab A (SM-T585)

### Android version

8.1.0

### Signal version

7.22.1

### Link to debug log

https://debuglogs.org/android/7.22.1/71fc8f447fce6ce9d3b86ba716626f972467575aa0e2b254814407dcbfd42285

**Comment by signal-support-request** [comment]
There is a deadlock detector in the logs complaining that all threads are stuck in `net.zetetic.database.sqlcipher.SQLiteConnectionPool.waitForConnection`.

Thread which seems to be not waiting for an SQ connection is `signal-JobRunner-1`:
```
-- [151] signal-JobRunner-1 | WAITING
java.lang.Object.wait(Native Method)
java.lang.Thread.parkFor$(Thread.java:2137)
sun.misc.Unsafe.park(Unsafe.java:358)
java.util.concurrent.locks.LockSupport.park(LockSupport.java:190)
java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:868)
java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(AbstractQueuedSynchronizer.java:900)
java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:1223)
java.util.concurrent.locks.ReentrantLock$NonfairSync.lock(ReentrantLock.java:210)
java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:286)
org.thoughtcrime.securesms.crypto.ReentrantSessionLock.acquire(ReentrantSessionLock.java:18)
org.thoughtcrime.securesms.crypto.storage.TextSecureSessionStore.containsSession(TextSecureSessionStore.java:77)
org.thoughtcrime.securesms.crypto.storage.SignalServiceAccountDataStoreImpl.containsSession(SignalServiceAccountDataStoreImpl.java:141)
org.whispersystems.signalservice.api.SignalServiceMessageSender.getEncryptedMessages(SignalServiceMessageSender.java:2634)
org.whispersystems.signalservice.api.SignalServiceMessageSender.sendMessage(SignalServiceMessageSender.java:1916)
org.whispersystems.signalservice.api.SignalServiceMessageSender.sendReceipt(SignalServiceMessageSender.java:240)
org.thoughtcrime.securesms.jobs.SendDeliveryReceiptJob.onRun(SendDeliveryReceiptJob.java:126)
org.thoughtcrime.securesms.jobs.BaseJob.run(BaseJob.java:31)
org.thoughtcrime.securesms.jobmanager.JobRunner.run(JobRunner.java:88)
org.thoughtcrime.securesms.jobmanager.JobRunner.run(JobRunner.java:51)
```

**Comment by greyson-signal** [comment]
Yeah, I see there's a lot of contention for the database lock for an extended period of time, and you're correct that there is a thread that's waiting on a different lock. Unfortunately I don't see how that'd create a deadlock in this instance, since that thread isn't holding the database lock at the time. So you'd think that eventually they'd resolve. And looking at the logs, the delivery receipts _do_ eventually send... after 25 min. So something is certainly going on.

**Comment by signal-support-request** [comment]
Probably something took this lock and then started a slow SQL query? I just don't see from the log which SQL query is currently executing, but there must at least be some thread that holds SQL as all other threads are locked waiting to get a connection.

**Comment by signal-support-request** [comment]
In the log there are lines like `DELETE FROM message WHERE thread_id = ? AND message.date_received < 1727477102671` saying the query has been running for hundreds of seconds. Apparently this query is very slow somehow, maybe missing index on both thread_id and date_received?

All this thread_id stuff was introduced recently, 83808ed427f27c102b76d5274ca421366a59a8ae is from Oct 25.

**Comment by greyson-signal** [comment]
That threadId stuff is unrelated to deleting, but good catch that there may be a very long-running delete. Deletes require having the database lock, and have known to be slow depending on the size, so that could very much be the issue. 

Unfortunately the reasons it's slow aren't index-related: it's all the related SQLite stuff that happens when you delete (triggers for FTS indices, foreign key checks, etc) that take lots of time. That said, we haven't seen it take _that_ long before, but could be that there's a series of deletes or something and they're all stacking up.

**Comment by signal-support-request** [comment]
> Those recent threadId changes are unrelated to deleting, but good catch that there may be a very long-running delete.

They may be unrelated, but in the log there is a query with `thread_id = ?`, so it should have a new index to speed it up.

**Comment by signal-support-request** [comment]
As a last resort option I can suggest that user goes to storage setting in Signal and click "Delete message history", then there will be no more messages to process (if this "delete everything" query is fast as it should be).

**Comment by signal-support-request** [comment]
The problem eventually fixed itself, after enabling VPN and disabling censorship circumvention. But it took a day to get fixed. Likely the same problem as #13750 indeed.
